### PR TITLE
Add a more complete app template for Flutter (skeleton)

### LINKED ISSUE #76379: A more complete app template for Flutter
## Use case

Flutter has one starter app template: the Counter app. It gets the user going quickly, and teaches low-level basics in 66 lines of code. We don't, however, provide a template that would be closer to a complete app, with several widgets in separate files, app-wide state, navigation or state restoration. 

This creates a large gap for novice users to leap over once they outgrow the one-file, one-widget, setState-only nature of Counter. And that leap lands them among a myriad of community solutions and opinions, most of which a novice user is not yet ready to review critically.

## Proposal

The goal is to make it easier for developers to continue their learning journey beyond the Counter app. A developer using Android Studio, for example, should be able to click on an icon in the "New Flutter Project" screen that gives her a more complete project to work with. 

This new template should co-exist with the original Counter app template. It is not a replacement.

## Reqiurements

  - Simplicity. It is understood that the new template will be less straightforward than the existing Counter app template. But even the new template SHOULD NOT be overwhelming to new Flutter developers, and it MUST NOT be overwhelming to experienced Flutter developers.
  - Teachability. The new template SHOULD be written in a way that allows gradual learning.
  - Accessibility. The new template app MUST be accessible. Screen reader navigation SHOULD be comfortable.
  - Beauty. The template app SHOULD look beautiful to the extent allowed by other provisions.
  - State management. The state management approach of the new template MUST make it simple for the developer to add new features in any part of the app.
  - Performance. The new template MUST follow basic performance guidelines, like avoiding unnecessary rebuilds.
  - Targets. The new template MUST support the following targets: mobile, desktop and web.

### LINKED ISSUE #84150: Run pub get --offline after flutter create
With the new `skeleton` template (https://github.com/flutter/flutter/pull/83530), we need to make sure we run `flutter pub get --offline` just after `flutter create`. This is because otherwise, `flutter run` will not work (see [my explanation](https://github.com/flutter/flutter/pull/83530#issuecomment-856202004)).

We already do run `flutter pub get` by default after project creation. https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/commands/create_base.dart#L423-L428
Looks like we might need to either wait a bit (some race condition?), or see if there's something else happening after this line that might mean we need to run it again.

I've volunteered to fix this.

### PR DESCRIPTION
**Pull request by filiph**
Fixes #76379.
Fixes #84150.

This is another stab at landing #76519, which was closed in March. The code is nearly identical, except:

* The template is now null safe
* The change is rebased to latest Flutter
* The change will depend on `flutter_template_images` version 4.0.0 (when https://github.com/flutter/packages/pull/366 lands)

The following is a direct copypaste from #76519, which was authored by @brianegan:

-----

## Summary

I've worked with @filiph to develop a new template application that includes two features: A static list / detail feature and a dynamic settings feature. When a user changes the Theme in the settings screen, the application responds in kind.

The template takes the following notable approaches. Reasons behind each feature / approach is documented in the ["New Flutter Template: Final Decisions" document.](https://docs.google.com/document/d/1sXAPfPaaaskbzRoA4fbAoZjohcd6jZXYYJJQd0lXcuA/edit?usp=sharing)

  - The template targets Intermediate Flutter developers. "Intermediate Flutter Developer" is defined by the "Flutter Learning Journey's" document.
  - Generally follows the [Simple App State Management guide](https://flutter.dev/docs/development/data-and-backend/state-mgmt/simple), using a ChangeNotifier to coordinate multiple Widgets.
  - Includes an analysis_options.yaml file
  - Generates localizations by default using arb files.
  - Includes an example image and establishes 1x, 2x and 3x folders for image assets.
  - Uses a "feature-first" folder organization
  - Navigator 1.0 with pushNamed for Navigation

## Approach

  1. Spent one week on research and competitive analysis. This involved recruiting and interviewing beginner to intermediate Flutter developers to determine their current learning targets and needs. Furthermore, I reviewed popular architecture and state management libraries from the Flutter community to determine the current advice and best practices offered by the Flutter ecosystem. Finally, I researched templates from "competing" frameworks, such as Android, iOS, Xamarin, React and Angular to see what similar frameworks offer in regards to templates. This research has been compiled into the ["New Flutter Template: Research and Competitive Analysis"](https://docs.google.com/document/d/10qG-NKoDrbozKzHW1JjaUHuKWor563p8JERNhkYlsVQ/edit?usp=sharing) Google doc that informed our subsequent work. Please note: This document includes some notes from the user interviews. It does not contain any personally revealing information, but to ensure privacy, I've restricted access to this document.
  2. Next, I spent 3 weeks coding up various approaches. This resulted in the creation of [~25 different potential templates](https://github.com/brianegan/new_flutter_template/tree/main/evaluations). From these evaluations we learned the different types of decisions we need to make for a final template. For example: What should the template do? Should it include localization? Logging? Responsive Breakpoints? Etc.
  3. We then asked for more [community feedback on these decisions](https://github.com/brianegan/new_flutter_template/issues). We used this feedback + all of the information gathered in our previous steps to make our final decisions. The decisions we've reached are documented in the ["New Flutter Template: Final Decisions" document.](https://docs.google.com/document/d/1sXAPfPaaaskbzRoA4fbAoZjohcd6jZXYYJJQd0lXcuA/edit?usp=sharing)
  4.  Now, we've taken those decisions and distilled them into this template proposal!

## Preview

https://user-images.githubusercontent.com/126604/108721724-a0ad9380-7522-11eb-97b9-44307757d93e.mp4

https://user-images.githubusercontent.com/126604/109329104-7d545280-785a-11eb-863b-561e7ac8c95d.mp4

---

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

### CODE REVIEW

**Comment by filiph** [comment]
This is ready for re-review. Some notes:

1. This is an updated version of https://github.com/flutter/flutter/pull/76519, which already got LGTM from @goderbauer, @zanderso, @Hixie, @csells.
2. Still, I'd like someone to go over the final product, including my changes:
    * The template is now null safe
    * The change is rebased to latest Flutter (~3 months of development)
3. The change will fail CI until I also land https://github.com/flutter/packages/pull/366. In the original PR, @stuartmorgan asks *not* to land the flutter/packages change until this one is fully approved and ready to land.

Also, @brianegan, can I ask you to  reply `@googlebot I consent.` here, please?

**Comment by csells** [comment]
lgtm

File: packages/flutter_tools/templates/skeleton/lib/src/app.dart.tmpl
Comment (goderbauer) [inline packages/flutter_tools/templates/skeleton/lib/src/app.dart.tmpl:37]: I think we also should also add `GlobalMaterialLocalizations.delegate` and `GlobalWidgetsLocalizations.delegate` here to avoid tripping users up if they start adding another language. (maybe even `GlobalCupertinoLocalizations.delegate` as well)

/cc @HansMuller

Comment (goderbauer) [reply packages/flutter_tools/templates/skeleton/lib/src/app.dart.tmpl:37]: Also: Should we explicitly specify the `supportedLocales` parameter here (and that it to its default `en` locale) to make users aware that if they want to support more languages, they'd have to change that list?

Comment (HansMuller) [reply packages/flutter_tools/templates/skeleton/lib/src/app.dart.tmpl:37]: Yes. Best to follow the guidelines in https://flutter.dev/docs/development/accessibility-and-localization/internationalization and point developers at that doc for more info.

**Comment by HansMuller** [comment]
Maybe [Dummy](https://www.google.com/search?q=dummy+synonym&oq=dummy+synonym&aqs=chrome..69i57j0l2j0i22i30l7.4889j0j7&sourceid=chrome&ie=UTF-8
) isn't the best prefix. Stub or proxy might be better.

File: packages/flutter_tools/lib/src/commands/create.dart
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/create.dart:45]: This should end with a period. If the tests didn't catch it, we should update the tests.

File: packages/flutter_tools/templates/skeleton/lib/src/dummy_feature/dummy_item.dart.tmpl
Comment (Hixie) [inline packages/flutter_tools/templates/skeleton/lib/src/dummy_feature/dummy_item.dart.tmpl:2]: Sample may be a better word than Dummy

File: packages/flutter_tools/templates/skeleton/pubspec.yaml.tmpl
Comment (Hixie) [inline packages/flutter_tools/templates/skeleton/pubspec.yaml.tmpl:32]: there are a lot fewer comments in this file than in the main template's pubspec.yaml, is that intended?

Comment (filiph) [reply packages/flutter_tools/templates/skeleton/pubspec.yaml.tmpl:32]: This is intentional: I thought it's okay since this template is focused on intermediate users who "just want to start an app". But I don't have a strong opinion here, and can easily revert back. Let me know.

> Aside: When the comments are back, 95% of the lines in this file are the same as in the main template (we're just actually using `assets:` here). I was looking at ways to make this DRY, but it would just make bind this new template to `create_base.dart` with something like [`withPluginHook`](https://github.com/flutter/flutter/blob/34059eec2c671472ee95f5093705fb9fed0b3919/packages/flutter_tools/lib/src/commands/create_base.dart#L368).

File: packages/flutter_tools/templates/skeleton/lib/src/app.dart.tmpl
Comment (filiph) [reply packages/flutter_tools/templates/skeleton/lib/src/app.dart.tmpl:37]: Ack. I'll add these.

File: packages/flutter_tools/lib/src/commands/create.dart
Comment (filiph) [reply packages/flutter_tools/lib/src/commands/create.dart]: Fixed.

Do you know which tests would catch this? I'm unable to run some of the tests locally, and due to the multi-package nature of this change, I can't see all the test results until 1) this CL is approved, 2) I land the `flutter/packages` change, 3) I re-run this tests on this change.

**Review by goderbauer — COMMENTED** [review]
The template LGTM

**Comment by filiph** [comment]
Renamed `Dummy*` to `Sample*`. Thanks for the naming idea, @Hixie, I really like it.

**Comment by brianegan** [comment]
@googlebot I consent.

**Comment by brianegan** [comment]
Looking good, glad to see this one moving forward :) Also, I like the change from Dummy to Sample as well :+1:

**Comment by filiph** [comment]
One complication that I only realized as I'm testing locally is the fact that adding generated localization from the very start invalidates our `flutter create` steps.

Observe:

```text
$ flutter create -t skeleton my_test
Creating project my_test...
  [...]
Wrote 133 files.

All done!
In order to run your application, type:

  $ cd my_test
  $ flutter run

Your application code is in my_test/lib/main.dart.
```

If you try to follow the instructions above, you get an error:

```text
$ cd my_test
$ flutter run
Launching lib/main.dart on Pixel 2 XL in debug mode...
Error: Couldn't resolve the package 'flutter_gen' in 'package:flutter_gen/gen_l10n/app_localizations.dart'.
lib/src/app.dart:2:8: Error: Not found: 'package:flutter_gen/gen_l10n/app_localizations.dart'
import 'package:flutter_gen/gen_l10n/app_localizations.dart';
       ^
[...]

Exception: Gradle task assembleDebug failed with exit code 1
```

What you need to do at that point is:

```text
$ flutter pub get
```

After that, `flutter run` works. This is confusing, since _some_ version of `flutter pub get` is called after `flutter create`, as far as I know. I guess it's running too soon? 

@jonahwilliams and I quickly chatted about this [on Discord](https://discord.com/channels/608014603317936148/608022056616853515/847601891969859604) because the same issue happens in testing. It's easy enough to run `flutter pub get --offline` in automated tests, but it's harder to ask users to remember to do that.

Now, I wonder if we want to fix the `flutter create` output (easier but uglier) or to start automatically doing `flutter pub get --offline` after create.

We never had this issue because no template had source generation in it from the very beginning.

**Comment by jonahwilliams** [comment]
> Now, I wonder if we want to fix the flutter create output (easier but uglier) or to start automatically doing flutter pub get --offline after create.

I think this sounds like a reasonable approach, it will be fast and we could always hide the output from that step too.

**Comment by filiph** [comment]
Ok, I'll add that to the CL, then.

**Comment by zanderso** [comment]
> Ok, I'll add that to the CL, then.

Should it go in a separate CL? This CL is already pretty big.

**Comment by filiph** [comment]
I'll be more than happy to do it in a separate one!

**Comment by filiph** [comment]
I filed the issue for the `flutter create` fix: https://github.com/flutter/flutter/issues/84150.

I consider this PR ready to go, so I asked @stuartmorgan for a review & merge of https://github.com/flutter/packages/pull/366, which should make the tests here go green.

**Comment by jonahwilliams** [comment]
All you need to do is set this line to true: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/commands/create_base.dart#L427

**Comment by jonahwilliams** [comment]
Sorry, totally slipped my mind

**Comment by filiph** [comment]
Sweet! Thanks @jonahwilliams, I made the change here.

Fixes #84150.

**Review by Hixie — APPROVED** [review]
LGTM except the tool changes for which please make sure @jonahwilliams has also given a positive review. Thanks.

**Review by jonahwilliams — APPROVED** [review]
Tool changes LGTM

You'll need to run flutter update-packages --force-upgrade to get the hash to update

**Comment by frankhasen** [comment]
I would be very happy if the template would implement the Navigator 2.

**Comment by anthonymoretti** [comment]
@filiph Should the `pubspec.yaml` also include the 2.0x and 3.0x image subdirectories?
```
  assets:
    # Add assets from the images directory to the application.
    - assets/images/
    - assets/images/2.0x/
    - assets/images/3.0x/
```
In [Specifying assets](https://flutter.dev/docs/development/ui/assets-and-images#specifying-assets) it says:
> To add files located in subdirectories, create an entry per directory.

**Comment by Add00w** [comment]
> I would be very happy if the template would implement the Navigator 2.

with bottom navigation too.

**Comment by stuartmorgan** [comment]
This PR has been closed since last year; the place to file new enhancement requests is the issue tracker.

**Pull request merged**
